### PR TITLE
Sanitize paste content by filtering out control chars

### PR DESCRIPTION
**Pull request by skyline75489**
## Summary of the Pull Request

Address the security issued mentioned in https://github.com/microsoft/terminal/issues/395#issuecomment-477206196

 
## References

Supersedes #7508

## PR Checklist
* [ ] Closes #xxx
* [ ] CLA signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/Terminal) and sign the CLA
* [ ] Tests added/passed
* [ ] Documentation updated. If checked, please file a pull request on [our docs repo](https://github.com/MicrosoftDocs/terminal) and link it here: #xxx
* [ ] Schema updated.
* [ ] I've discussed this with core contributors already. If not checked, I'm ready to accept this work might be rejected in favor of a different grand plan. Issue number where discussion took place: #xxx

### CODE REVIEW

**Comment by zadjii-msft** [comment]
@msftbot make sure @DHowett signs off on this

**Review by zadjii-msft — APPROVED** [review]
This seems reasonable to me, but I want Dustin to take a look too, since he's the expert in that particular area.

File: src/cascadia/TerminalControl/TermControl.cpp [hunk 1]
```diff
@@ -2038,6 +2038,27 @@ namespace winrt::Microsoft::Terminal::TerminalControl::implementation
...
+                // Not a control character.
+                return false;
+            }
+
+            return c >= L'\x00' && c <= L'\x08' ||
+                   c >= L'\x0b' && c <= L'\x0c' ||
+                   c >= L'\x0e' && c <= L'\x1f';
+        }));
+
         // Some notes on this implementation:
```
Comment (DHowett) [inline src/cascadia/TerminalControl/TermControl.cpp:2062]: i'd rather we not scan the string multiple times. can we consolidate this with the loop below?

**Review by DHowett — CHANGES_REQUESTED** [review]
(making my NAK an actual "request changes")

**Comment by skyline75489** [comment]
I should probably mention that thanks to #8744, the potentially malicious content from the clipboard is being displayed in a dazzling dialog. This *should* prevent users from pasting it in most conditions.

image

**Comment by skyline75489** [comment]
This PR will still help prevent attacks described in https://security.stackexchange.com/q/39118, where currently the dialog isn't showing invisible control character.s

image

File: src/cascadia/TerminalControl/TermControl.cpp [hunk 2]
```diff
@@ -2049,6 +2049,33 @@ namespace winrt::Microsoft::Terminal::TerminalControl::implementation
         //   the \n (but not including it). Then, we check the if the
         //   previous character is \r, if its not, then we had a lone \n
         //   and so we append our own \r
+        // - For security reasons, most control characters should be removed.
+
+        auto pred = [](wchar_t c) {
```
Comment (skyline75489) [inline src/cascadia/TerminalControl/TermControl.cpp:2054]: Inlining this in `std::remove_if` results in very weird indention using `clang-format`. Nesting this inside `sanitize` crashes `clang-format`(!). So I had to put it here.

File: src/cascadia/TerminalControl/TermControl.cpp [hunk 3]
```diff
@@ -2049,6 +2049,33 @@ namespace winrt::Microsoft::Terminal::TerminalControl::implementation
...
+            }
+
+            if (c > L'\x7f')
+            {
+                // Not a control character.
+                return false;
+            }
+
+            // All ASCII control characters will be removed except HT(0x09), LF(0x0a),
+            // CR(0x0d) and DEL(0x7F).
```
Comment (skyline75489) [inline src/cascadia/TerminalControl/TermControl.cpp:2068]: Removing `CR` and `LF` seems meaningless from a practical point of view. But generally I think this function *should* remove `CR` and `LF` had it not been in this particular context.

Comment (skyline75489) [reply src/cascadia/TerminalControl/TermControl.cpp:2068]: (copied from #7508)

The best resource I can find is here https://security.stackexchange.com/a/52655. It suggest removing every control characters except:

* CR(0x0d) \r 
* LF(0x0a) \n 
* HT(0x09) \t, 
* DEL(0x7F)

Then there's some other opinions in https://bugzilla.gnome.org/show_bug.cgi?id=753197, regarding removing:

* 0x80-0x9F

And here https://bugzilla.gnome.org/show_bug.cgi?id=794653, regarding removing:

* BS(0x08)
* DEL(0x7F)

**Comment by jantari** [comment]
I just want to say I do intentionally paste text with VT control sequences into the terminal, usually just for colors or other effects. So I'm not sure how I feel about `0x1b` aka `ESC` being stripped

**Comment by DHowett** [comment]
I suspect you are in the minority here. Many other terminal applications filter out control characters during bracketed paste, and _an application shouldn't be expecting to receive raw SGRs in its input stream anyway_.

**Comment by jantari** [comment]
Yea I understand that, but what about an option - either in the JSON config, or directly in the new paste dialog:

paste-confirmation

When it detects non-printable or control-characters in the text, it could present a third, non-focused button: "Paste with special characters"

**Comment by DHowett** [comment]
It's just that... every setting we add is another combinatorial factor in our testing, right? There's already `sendInput` for sending raw controlled input. I just don't see the use case for pasting control characters today..?

**Comment by jantari** [comment]
Good point, possible usecases were testing string snippets for shell prompts (bash `$PS1` or powershells' `prompt` function) or other snippets of fancy text. It's not a big issue though, by now I've more or less switched entirely to shell-sequences like `\033` in bash or the ugly `$([char]27)` in powershell to get `ESC` rather than a raw `ESC` character

Since you said `SendInput` is not affected by this behavior/change - I assume the same holds true for `WriteConsoleInput`?

**Comment by DHowett** [comment]
The code change in this file, and the policy we're instating here, implicates **only** paste.

We're not breaking Win32 APIs willy-nilly. :smile:

Also: PowerShell Core supports "`e" for escaping escape.

**Comment by skyline75489** [comment]
I am no expert on security issue like this. I think this PR needs to be thoroughly reviewed.  Feel free to tag this to 1.7 or future release. No need to rush this one.

**Comment by skyline75489** [comment]
I *think* I should also add this to conhost. But I don't know conhost that much. I'll file a issue for conhost if this is merged.

**Comment by DHowett** [comment]
I appreciate this, but I do think that it needs some changes.

I'm not a huge fan of the lambda filtering implementation; I think that it would be much better for us to have _one loop_ that checks for `\n` _and all the other control sequences at the same time._

As it is implemented in this PR, we scan the string to determine that it doesn't contain `\n`, and then we re-scan the portion between 0 and the first \n, then we re-scan the portion between that \n and the next one, and then ... (etc) to detect control sequences.

Same concern about the bracketed paste function. Why have a lambda?

Lastly- I wonder if this should be inside TerminalCore so that the WPF control can use it. Needing the correct filtering for \r\n _and_ for control sequences is not something unique to the TerminalControl!

**Comment by skyline75489** [comment]
The main reason I used lambda is to minimize the modification to the existing code. Yea it does introduced extra scan loop. I can’t blame you for wanting this to run faster. I’ll try to refactor  this later.

**Comment by skyline75489** [comment]
Closed in favor of #9034

**Pull request closed without merging**